Document InsertTagParser service (#1154)
contao/docs: docs/dev/reference/services.md

}
```

## InsertTagParser

{{< version "4.13" >}}

This service let you replace inserttags in your string content.

```php
use Contao\CoreBundle\InsertTag\ChunkedText;
use Contao\CoreBundle\InsertTag\InsertTagParser;

class Example
{
    private InsertTagParser $insertTagParser;

    public function __invoke(string $buffer): string
    {

>         // Return a string that may contain esi tags

Reviewer: I think many people won't understand ESI. I don't think it's super helpful to describe what the methods do. It would be way more valuable to document **when** to use which method. So real world examples.

- `replace()` should be used in HTML context when the replaced result is sent as a response to the client. It will allow for ESI tags that can improve caching behaviour.
- `replaceInline()` should be used when the result is **not** sent to a client. Like e.g. when you use `$email->subject()` or so.

The chunked alternatives have very specific use cases which I think 99% of the developers will never need. But maybe @m-vo or @ausi can come up with a real world example here too?

Author: The example for chunked texts was taken from the twig ContaoExtension, I just replace twig_escape_filter with htmlspecialchars :)
I will adjust the texts, sound good :+1:

}
```

## InsertTagParser

{{< version "4.13" >}}

This service let you replace inserttags in your string content.

Some methods return a `ChunkedText` instance. The `ChunkedText` container was created to keep the surrounding text containing the insert tags separate from the replacements made by the insert tag parser. It is used for example in the twig escaper to skip encoding on inserttag replacement results.

```php
use Contao\CoreBundle\InsertTag\ChunkedText;
use Contao\CoreBundle\InsertTag\InsertTagParser;

class Example
{
    private InsertTagParser $insertTagParser;

    public function __invoke(string $buffer): string
    {

>         // Return a string and should be used in HTML context when the replaced result is sent as a response to the client. It will allow for ESI tags that can improve caching behaviour.

Reviewer 2: ```suggestion
        // Return a string and should be used in HTML context when the replaced result is sent as a response to the 
        // client. It will allow for ESI tags that can improve caching behaviour.
```

        $resultReplace = $this->insertTagParser->replace($buffer);

>         // Return a string and should be used when the result is not sent to a client ( e.g. when using `$email->subject()`).

Reviewer 2: ```suggestion
        // Return a string and should be used when the result is not sent to a client 
        // (e.g. when using `$email->subject()`).
```

        $resultReplaceInline = $this->insertTagParser->replaceInline($buffer);

>         // Return ChunkedText instance and should be used in HTML context when the replaced result is sent as a response to the client. It will allow for ESI tags that can improve caching behaviour.

Reviewer 2: ```suggestion
        // Return ChunkedText instance and should be used in HTML context when the replaced result is sent as a 
        // response to the client. It will allow for ESI tags that can improve caching behaviour.
```

}
```

## InsertTagParser

{{< version "4.13" >}}

> This service let you replace inserttags in your string content.

Reviewer: ```suggestion
This service lets you replace insert tags in your string content.
```

Some methods return a `ChunkedText` instance. The `ChunkedText` container was created to keep the surrounding text 
containing the insert tags separate from the replacements made by the insert tag parser. It is used for example in the 

> twig escaper to skip encoding on inserttag replacement results.

Reviewer: ```suggestion
twig escaper to skip encoding on insert tag replacement results.
```
